jib-core/src/main/java/com/google/cloud/tools/jib/cache/CacheStorageReader.java

```diff
@@ -81,6 +81,38 @@ static void verifyImageMetadata(ImageMetadataTemplate metadata, Path metadataCac
     this.cacheStorageFiles = cacheStorageFiles;
   }
 
+  /**
+   * Returns {@code true} if all image layers described in a manifest have a corresponding file
+   * entry in the cache.
+   *
+   * @param manifest the image manifest
+   * @return a boolean
+   */
+  boolean areAllLayersCached(ManifestTemplate manifest) {
```

Comment:
@mpeddada1 For multi-module parallel builds, should this method or the part that deals with the filesystem be `synchronized`? I wonder what will happen if two different threads have a race condition and are both in this method at the same time.

Reply:
Jib has two approaches for synchronization in different parts: 1) using atomic move (renaming) and 2) explicit locking with a lock file. (I was never a big fan of the first approach despite its simplicity, which had issues like https://github.com/GoogleContainerTools/jib/issues/3740#issuecomment-1225967476 and #2318. That's why we started using the second approach for stuff like caching the metadata). Here, I believe layer directories are set up via atomic move, although you'll want to double-check. So whenever you start to see a layer directory, then it is there, and you're good to go. (There may be concurrent atomic moves to set up the directory, but they will silently fail once the directory is already set up by someone else.)

Note that even if we assume that it needed synchronization, `synchronized` would help only with a single Jib process. The cache should work correctly even if there are multiple JVMs running Jib.

Reply:
@elefeint great point, you bring up something that I didn't think of before in terms of sufficient handling of concurrency. 

@chanseokoh thanks for adding more context to this! You're right, @emmileaf discovered that [we do maintain atomicity of writes](https://github.com/GoogleContainerTools/jib/pull/3767#issuecomment-1242003718) when populating a cache. But similar to what you've mentioned, [there are also places that warn of insufficient synchronization](https://effectivejava.googleplex.com/third-edition#page=342).  Thinking out loud/two things that I'm curious about:

1) Noticed a [proposal document](https://github.com/GoogleContainerTools/jib/blob/0ed7dca36864b6b19ff61629fc578018041fa15f/proposals/archives/cache_v2.md#locking) which mentions that using FileLocks can result in a `OverlappingFileLockException` during concurrent runs of Jib within the same JVM. Could this be a potential risk?
2) Would synchronizing writes and reads be a major overhaul?

Reply:
Sorry, I wasn't saying there's any place of insufficient synchronization. It's the opposite; with the atomic move, there's no concurrency issue. (It's only that a Jib run may fail when an atomic move is blocked or not supported for any reasons. Even so, there's no data corruption.)

The proposal is an old one, and it may be that the author did not have a full understanding of `OverlappingFileLockException`. You just shouldn't do concurrent access within the same JVM, which is what [Jib is currently handling correctly](https://github.com/GoogleContainerTools/jib/blob/8004ec6f5ac21783b5121b66c4d132b95118d2c8/jib-core/src/main/java/com/google/cloud/tools/jib/filesystem/LockFile.java#L55-L57).

Reply:
OH sorry I totally misread the comment lol. 

> It's only that a Jib run may fail when an atomic move is blocked or not supported for any reasons. Even so, there's no data corruption.

Got it, that makes sense. Hence, the instance where we had to improve our error messaging to cater for filesystem specific issues with atomic moves (e.g. antivirus).